ministryofjustice/correspondence_tool_staff
# Ct 4128 implement off sars closure logic

## Description
This PR updates the case closure logic for Offender SAR and Offender SAR Complaint cases so that as they are closed the case in question and its related cases receives a RetentionSchedule object which is used to mark a case for GRPR destruction etc.

For **Offender SARs** the way that cases receive retention schedules is like the following:

![offendersarcases](https://user-images.githubusercontent.com/13377553/166891116-00328288-674c-4720-9c9f-54c851c7bf91.jpg)

Each receives a separate instance of RetentionSchedule with the same `planned_erasure_date` as the initial case that was closed.

For **Offender SAR Complaints** it is slightly more complicated:

- RS = RetentionSchedule

![offsarcomplaints](https://user-images.githubusercontent.com/13377553/166892820-d7979fb0-2f30-4c23-b39a-44c357b5edd0.jpg)

When the complaint case is closed it receives a RetentionSchedule, and so does it's original case (i.e. the case that the initial complaint is about).

However, the original case can have more than one complaint against it. So each of these "other" complaints against the case also gets a RetentionSchedule with the same date as the initial complaint case, if each case is in a closed state.  

## Self-review checklist



### Screenshots
There are not many UI changes - only that the planned erasure date is now visible on the the show pages for each case:

<img width="982" alt="Screenshot 2022-05-05 at 10 37 06" src="https://user-images.githubusercontent.com/13377553/166897842-9631048e-79d6-4025-90be-1ccfd3ed7be8.png">


### Related JIRA tickets
[CT-4128](https://dsdmoj.atlassian.net/browse/CT-4128)
[CT-4129](https://dsdmoj.atlassian.net/browse/CT-4129)

### Deployment


### Manual testing instructions
For Offender SARs:
- Create some Offender SARs and link them, take one to closure, after one is closed both should have a planned erasure date on their separate show pages.

For Offender SAR complaints:
- Create a complaint case against an existing offender SAR case 
- Log into the console and find amend its created_at date to some time a few days ago (this is needed so that we can spot the difference in updates between different complaint cases later on. 
- then take it through to closure. 
- You should see a planned erasure date on both the Complaint Case, and Offender SAR cases' show page.
- Create another Complaint against the same Offender SAR case, and keep it open
- Create another Complaint against the same Offender SAR case and take it through to closure, once closed the planned_erasure_date on the first complaint case we created, and the Offender SAR case the complaint was about should all be the same on their respective show pages.
- However the open Complaint should not have a plan erasure date on its show page at all.
